File: oxidize-golang/core/model/llama_decoder.go
package model

// LlamaDecoderConfig holds tensor geometry for a standard Llama-style decoder stack.
type LlamaDecoderConfig struct {
	HiddenSize          int
	LayerCount          int
	IntermediateSize    int
	NumAttentionHeads   int
	NumKeyValueHeads    int
	KeyValueHeadDim     int
	VocabSize           int
	RMSNormEps          float32
	RopeTheta           float32
	NumExperts          int
	NumExpertsPerToken  int
	SlidingWindow       int
	UseAlibi            bool
	ParallelAttnFFN     bool
	// Gemma-family fields.
	RopeThetaSWA         float32
	SlidingWindowPattern int
	EmbeddingScale       float32
	GeluFFN              bool
	SandwichNorm         bool
}

// LayerIsGlobal reports whether layer idx uses global (full-context) attention
func (c LlamaDecoderConfig) LayerIsGlobal(idx int) bool {
	if c.SlidingWindow == 0 {
		return true
	}
	if c.SlidingWindowPattern == 0 {
		return false
	}
	return (idx+1)%c.SlidingWindowPattern == 0
}

// LayerRopeTheta returns the RoPE base for layer idx: global layers use
// RopeTheta; sliding-window layers use RopeThetaSWA when set.
func (c LlamaDecoderConfig) LayerRopeTheta(idx int) float32 {
	if c.RopeThetaSWA > 0 && !c.LayerIsGlobal(idx) {
		return c.RopeThetaSWA
	}
	return c.RopeTheta
}

// LayerSlidingWindow returns the effective window for layer idx (0 = full).
func (c LlamaDecoderConfig) LayerSlidingWindow(idx int) int {
	if c.SlidingWindow > 0 && !c.LayerIsGlobal(idx) {
		return c.SlidingWindow
	}
	return 0
}

// HeadDim returns hidden_size / num_attention_heads.
func (c LlamaDecoderConfig) HeadDim() int {
	if c.NumAttentionHeads == 0 {
		return 0
	}
	return c.HiddenSize / c.NumAttentionHeads
}

// KVHeadDim returns per-head K/V dimension.
func (c LlamaDecoderConfig) KVHeadDim() int {
	if c.KeyValueHeadDim > 0 {
		return c.KeyValueHeadDim
	}
	return c.HeadDim()
}

// LlamaDecoderConfigFromInference maps InferenceConfig into decoder geometry.
func LlamaDecoderConfigFromInference(cfg InferenceConfig) LlamaDecoderConfig {
	window := cfg.SlidingWindow
	if window == 0 && cfg.Architecture.UsesSlidingWindow() {
		window = 4096
	}
	return LlamaDecoderConfig{
		HiddenSize:         cfg.HiddenSize,
		LayerCount:         cfg.LayerCount,
		IntermediateSize:   cfg.IntermediateSize,
		NumAttentionHeads:    cfg.NumAttentionHeads,
		NumKeyValueHeads:   cfg.NumKeyValueHeads,
		KeyValueHeadDim:    cfg.KVHeadDim(),
		VocabSize:          cfg.VocabSize,
		RMSNormEps:         cfg.RMSNormEps,
		RopeTheta:          cfg.RopeTheta,
		NumExperts:         cfg.NumExperts,
		NumExpertsPerToken: cfg.NumExpertsPerToken,
		SlidingWindow:      window,
		UseAlibi:           cfg.Architecture.UsesAlibi(),
		ParallelAttnFFN:    cfg.Architecture.UsesParallelAttnFFN(),
		RopeThetaSWA:         cfg.RopeThetaSWA,
		SlidingWindowPattern: cfg.SlidingWindowPattern,
		EmbeddingScale:       cfg.EmbeddingScale,
		GeluFFN:              cfg.GeluFFN,
		SandwichNorm:         cfg.SandwichNorm,
	}
}

// LlamaDecoderConfigFromDFlash maps DFlashConfig into decoder geometry.
func LlamaDecoderConfigFromDFlash(cfg DFlashConfig) LlamaDecoderConfig {
	return LlamaDecoderConfig{
		HiddenSize:        cfg.HiddenSize,
		LayerCount:        cfg.NumHiddenLayers,
		IntermediateSize:  cfg.IntermediateSize,
		NumAttentionHeads: cfg.NumAttentionHeads,
		NumKeyValueHeads:  cfg.NumKeyValueHeads,
		VocabSize:         cfg.VocabSize,
		RMSNormEps:        cfg.RMSNormEps,
		RopeTheta:         cfg.RopeTheta,
	}
}

// DecoderAttentionLayer holds per-layer attention projections.
type DecoderAttentionLayer struct {
	QProj, KProj, VProj, OProj F32Weight
	QNormWeight, KNormWeight   []float32
}

// DecoderLayer is norm → attention → MLP (dense and/or MoE).
type DecoderLayer struct {
	InputLayernorm, PostAttentionLayernorm []float32
	PreFFNLayernorm, PostFFNLayernorm      []float32
	Attention                              DecoderAttentionLayer
	MLPGate, MLPUp, MLPDown                F32Weight
	MoE                                    MoELayer
}

// DecoderKvLayerCache stores per-layer KV in flattened [seq][kv_heads*head_dim] layout.
type DecoderKvLayerCache struct {
	Keys, Values []float32
	SeqLen       int
}

// DFlashAttentionLayer aliases DecoderAttentionLayer for the DFlash draft path.
type DFlashAttentionLayer = DecoderAttentionLayer

// DFlashDecoderLayer aliases DecoderLayer for the DFlash draft path.
type DFlashDecoderLayer = DecoderLayer

// DFlashKvLayerCache aliases DecoderKvLayerCache for the DFlash draft path.
type DFlashKvLayerCache = DecoderKvLayerCache

// NewDFlashKvLayerCache returns an empty cache.
func NewDFlashKvLayerCache() DFlashKvLayerCache { return DecoderKvLayerCache{} }

func (c *DecoderKvLayerCache) clear() {
	c.Keys = c.Keys[:0]
	c.Values = c.Values[:0]
	c.SeqLen = 0
}

func (c *DecoderKvLayerCache) reserveTokens(additional, kvLen int) {
	add := additional * kvLen
	if cap(c.Keys)-len(c.Keys) < add {
		c.Keys = append(c.Keys, make([]float32, 0, len(c.Keys)+add)...)
	}
	if cap(c.Values)-len(c.Values) < add {
		c.Values = append(c.Values, make([]float32, 0, len(c.Values)+add)...)
	}
}

// LlamaDecoderStack is the shared Llama-style transformer body used by InferenceModel
// and DFlashDraftModel.
type LlamaDecoderStack struct {
	Config         LlamaDecoderConfig
	Layers         []DecoderLayer
	TokEmbeddings  F32Weight
	Output         F32Weight
	Norm           []float32
	KVCache        []DecoderKvLayerCache
	PositionOffset int
	AlibiSlopes    []float32
}

// NewLlamaDecoderStack returns an empty stack sized for config.LayerCount layers.
func NewLlamaDecoderStack(config LlamaDecoderConfig) *LlamaDecoderStack {
	s := &LlamaDecoderStack{
		Config:  config,
		KVCache: make([]DecoderKvLayerCache, config.LayerCount),
	}
	if config.UseAlibi && config.NumAttentionHeads > 0 {
		s.AlibiSlopes = AlibiSlopes(config.NumAttentionHeads)
	}
	return s
}

// Loaded reports whether weights required for forward are present.
func (s *LlamaDecoderStack) Loaded() bool {
	return s.TokEmbeddings.IsLoaded() && s.Output.IsLoaded()
}

// ResetCache clears KV cache and position.
func (s *LlamaDecoderStack) ResetCache() {
	if len(s.KVCache) != s.Config.LayerCount {
		s.KVCache = make([]DecoderKvLayerCache, s.Config.LayerCount)
	} else {
		for i := range s.KVCache {
			s.KVCache[i].clear()
		}
	}
	s.PositionOffset = 0
}

// ReserveCacheTokens pre-allocates KV cache capacity.
func (s *LlamaDecoderStack) ReserveCacheTokens(tokens int) {
	kvLen := s.Config.NumKeyValueHeads * s.Config.KVHeadDim()
	for i := range s.KVCache {
		s.KVCache[i].reserveTokens(tokens, kvLen)
	}
}

// scaleEmbedding multiplies the embedding by Config.EmbeddingScale (Gemma uses
// sqrt(hidden_size)); a scale of 0 or 1 is a no-op.
func (s *LlamaDecoderStack) scaleEmbedding(x []float32) {
	scale := s.Config.EmbeddingScale
	if scale == 0 || scale == 1 {
		return
	}
	for i := range x {
		x[i] *= scale
	}
}

func (s *LlamaDecoderStack) fillTokenEmbedding(token uint32, output []float32) error {
	if !s.TokEmbeddings.IsLoaded() {
		return nil
	}
	vocab := s.TokEmbeddings.outputDim()
	if vocab < 1 {
		vocab = 1
	}
	idx := int(token)
	if idx >= vocab {
		idx = vocab - 1
	}
	embW := s.TokEmbeddings.inputDim()
	if embW == len(output) {
		return s.TokEmbeddings.Row(idx, output)
	}
	emb := make([]float32, embW)
	if err := s.TokEmbeddings.Row(idx, emb); err != nil {
		return err
	}
	n := len(output)
	if n > len(emb) {
		n = len(emb)
	}
	copy(output[:n], emb[:n])
	return nil
}

File: oxidize-python/oxidize_python/core/model/llama_decoder.py
"""Llama decoder stack types mirroring oxidize-golang/core/model/llama_decoder.go."""

from __future__ import annotations

from dataclasses import dataclass, field

from oxidize_python.core.model.dflash_weights import F32Weight


@dataclass
class LlamaDecoderConfig:
    hidden_size: int
    layer_count: int
    intermediate_size: int
    num_attention_heads: int
    num_key_value_heads: int
    key_value_head_dim: int
    vocab_size: int
    rms_norm_eps: float
    rope_theta: float
    # Gemma-family fields.
    sliding_window: int = 0
    rope_theta_swa: float = 0.0
    sliding_window_pattern: int = 0
    embedding_scale: float = 1.0
    gelu_ffn: bool = False
    sandwich_norm: bool = False

    def head_dim(self) -> int:
        if self.num_attention_heads == 0:
            return 0
        return self.hidden_size // self.num_attention_heads

    def kv_head_dim(self) -> int:
        if self.key_value_head_dim > 0:
            return self.key_value_head_dim
        return self.head_dim()

    def layer_is_global(self, idx: int) -> bool:
        """Whether layer idx uses global (full) attention rather than SWA.

        No window -> global. Window but no interleave pattern (Mistral) ->
        local. Gemma: every Nth layer (1-indexed) is global.
        """
        if self.sliding_window == 0:
            return True
        if self.sliding_window_pattern == 0:
            return False
        return (idx + 1) % self.sliding_window_pattern == 0

    def layer_rope_theta(self, idx: int) -> float:
        if self.rope_theta_swa > 0 and not self.layer_is_global(idx):
            return self.rope_theta_swa
        return self.rope_theta

    def layer_sliding_window(self, idx: int) -> int:
        if self.sliding_window > 0 and not self.layer_is_global(idx):
            return self.sliding_window
        return 0


@dataclass
class DecoderAttentionLayer:
    q_proj: F32Weight = field(default_factory=F32Weight.empty)
    k_proj: F32Weight = field(default_factory=F32Weight.empty)
    v_proj: F32Weight = field(default_factory=F32Weight.empty)
    o_proj: F32Weight = field(default_factory=F32Weight.empty)
    q_norm_weight: list[float] = field(default_factory=list)
    k_norm_weight: list[float] = field(default_factory=list)


@dataclass
class DecoderLayer:
    input_layernorm: list[float] = field(default_factory=list)
    # For Gemma (sandwich norm) post_attention_layernorm is the true
    # post-attention norm, pre_ffn_layernorm is the pre-MLP norm, and
    # post_ffn_layernorm is applied to the FFN output before its residual.
    post_attention_layernorm: list[float] = field(default_factory=list)
    pre_ffn_layernorm: list[float] = field(default_factory=list)
    post_ffn_layernorm: list[float] = field(default_factory=list)
    attention: DecoderAttentionLayer = field(default_factory=DecoderAttentionLayer)
    mlp_gate: F32Weight = field(default_factory=F32Weight.empty)
    mlp_up: F32Weight = field(default_factory=F32Weight.empty)
    mlp_down: F32Weight = field(default_factory=F32Weight.empty)


@dataclass
class DecoderKvLayerCache:
    keys: list[float] = field(default_factory=list)
    values: list[float] = field(default_factory=list)
    seq_len: int = 0

    def clear(self) -> None:
        self.keys = []
        self.values = []
        self.seq_len = 0

    def reserve_tokens(self, additional: int, kv_len: int) -> None:
        add = additional * kv_len
        if len(self.keys) + add > len(self.keys):
            self.keys.extend([0.0] * max(0, add))
        if len(self.values) + add > len(self.values):
            self.values.extend([0.0] * max(0, add))


@dataclass
class LlamaDecoderStack:
    config: LlamaDecoderConfig
    layers: list[DecoderLayer] = field(default_factory=list)
    tok_embeddings: F32Weight = field(default_factory=F32Weight.empty)
    output: F32Weight = field(default_factory=F32Weight.empty)
    norm: list[float] = field(default_factory=list)
    kv_cache: list[DecoderKvLayerCache] = field(default_factory=list)
    position_offset: int = 0

    @staticmethod
    def new(config: LlamaDecoderConfig) -> LlamaDecoderStack:
        return LlamaDecoderStack(
            config=config,
            kv_cache=[DecoderKvLayerCache() for _ in range(config.layer_count)],
        )

    def loaded(self) -> bool:
        return self.tok_embeddings.is_loaded() and self.output.is_loaded()

    def reset_cache(self) -> None:
        if len(self.kv_cache) != self.config.layer_count:
            self.kv_cache = [DecoderKvLayerCache() for _ in range(self.config.layer_count)]
        else:
            for c in self.kv_cache:
                c.clear()
        self.position_offset = 0

    def reserve_cache_tokens(self, tokens: int) -> None:
        kv_len = self.config.num_key_value_heads * self.config.kv_head_dim()
        for c in self.kv_cache:
            c.reserve_tokens(tokens, kv_len)

    def scale_embedding(self, x: list[float]) -> None:
        """Multiply embedding by config.embedding_scale (Gemma uses sqrt(hidden))."""
        scale = self.config.embedding_scale
        if scale in (0.0, 1.0):
            return
        for i in range(len(x)):
            x[i] *= scale

    def fill_token_embedding(self, token: int, output: list[float]) -> None:
        if not self.tok_embeddings.is_loaded():
            return
        vocab = self.tok_embeddings._output_dim()
        if vocab < 1:
            vocab = 1
        idx = token
        if idx >= vocab:
            idx = vocab - 1
        emb_w = self.tok_embeddings._input_dim()
        if emb_w == len(output):
            self.tok_embeddings.row(idx, output)
            return
        emb = [0.0] * emb_w
        self.tok_embeddings.row(idx, emb)
        n = min(len(output), len(emb))
        output[:n] = emb[:n]
